pyanaconda/modules/payloads/payload/flatpak/source.py

```diff
@@ -0,0 +1,435 @@
+        if self._is_local:
+            return "oci:" + self._url.removeprefix("file://")
+
+        collection_location = os.path.join(download_location, "Flatpaks")
+        expanded_refs = self._expand_refs(refs)
+
+        index_json = {
+            "schemaVersion": 2,
+            "manifests": []
+        }
+
+        with self._downloader() as downloader:
+            for image in self._images:
+                if image.ref in expanded_refs:
+                    log.debug("Downloading %s, %s bytes", image.ref, image.download_size)
+                    if progress:
+                        progress.report_progress(_("Downloading {}").format(image.ref))
+
+                    manifest_len = self._download_blob(downloader, download_location, image.digest)
+                    self._download_blob(downloader,
+                                        download_location, image.manifest_json["config"]["digest"])
+                    index_json["manifests"].append({
+                        "mediaType": "application/vnd.oci.image.manifest.v1+json",
+                        "digest": image.digest,
+                        "size": manifest_len
+                    })
+
+                    for layer in image.manifest_json["layers"]:
+                        self._download_blob(downloader,
+                                            download_location, layer["digest"],
+                                            stream=True)
+
+        os.makedirs(collection_location, exist_ok=True)
+        with open(os.path.join(collection_location, "index.json"), "w") as f:
+            json.dump(index_json, f)
+
+        with open(os.path.join(collection_location, "oci-layout"), "w") as f:
+            json.dump({
+                "imageLayoutVersion": "1.0.0"
+            }, f)
+
+        return "oci:" + collection_location
+
+    @cached_property
+    def _images(self) -> List[StaticSourceImage]:
+        result = []
+
+        with self._downloader() as downloader:
+            url = self._url + "/index.json"
+            response = downloader(url)
+            if response.status_code == 404:
+                raise NoSourceError("No Flatpak source found at {}".format(url))
+            response.raise_for_status()
+            index_json = response.json()
+
+            for manifest in index_json.get("manifests", ()):
+                if manifest.get("mediaType") == "application/vnd.oci.image.manifest.v1+json":
+                    digest = manifest["digest"]
+                    manifest_json = self._get_json(downloader, manifest["digest"])
+                    config_json = self._get_json(downloader, manifest_json["config"]["digest"])
+                    result.append(StaticSourceImage(digest, manifest_json, config_json))
+
+        return result
+
+    def _blob_url(self, digest):
+        if not digest.startswith("sha256:"):
+            raise RuntimeError("Only SHA-256 digests are supported")
+        return self._url + "/blobs/sha256/" + digest[7:]
```

Comment:
This is really low level & should at least point to a specification that clearly shows what elements there are and why we are fine with the ones after this index. :P